pkg/api/types.go

```diff
@@ -1363,6 +1364,23 @@ func (k *KubernetesConfig) GetAzureCNIURLWindows(cloudSpecConfig AzureEnvironmen
 	return cloudSpecConfig.KubernetesSpecConfig.VnetCNIWindowsPluginsDownloadURL
 }
 
+// IsFeatureEnabled returns true if a feature flag is on for the provided feature
+func (f *FeatureFlags) IsFeatureEnabled(feature string) bool {
+	if f != nil {
```

Comment:
@tariq1890 thoughts on nil checking a pointer self instance reference?

Reply:
I prefer nil check before the method call.

Also, if we ensure that a FeatureFlag struct is always initialized while instantiating the properties struct, then we can do away with the nil check.

Reply:
Me too. I'm not convinced we all think this is an improvement in quality.

Reply:
@CecileRobertMichon Correct me if I am wrong, but nil check within the function ensures that the field of a nil pointer is never accessed  but the method call still happens and in this case, the first if condition fails and goes straight to false.

I am suggesting that we do the nil check outside of that method  and before the method is invoked. This pattern would stay consistent when we try to safely access both sub-fields and methods from a struct field.

Reply:
> nil check within the function ensures that the field of a nil pointer is never accessed but the method call still happens and in this case, the first if condition fails and goes straight to false.

This is exactly right, I've updated my previous comment to be more clear.

My argument was that checking for nil outside the function means 1) duplicated nil checks if we call the same function multiple times and 2) prone to bugs if that function is ever called without a nil check 3) we can't unit test that the right behavior happens when the struct is nil (ie. that we return false)

Reply:
This restructuring of the code by taking out the function of template_generator and unit testing it was a prototype, I was thinking of changing the other functions to follow the same pattern in a later PR so let's get it right and exclude  consistency from reasoning.

Reply:
I'm happy (1) not setting the bar as high as getting it right for all time and (2) not being 100% consistent.

/lgtm